eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala

```diff
@@ -807,25 +815,36 @@ object Router {
     * @param numRoutes    the number of shortest-paths to find
     * @param extraEdges   a set of extra edges we want to CONSIDER during the search
     * @param ignoredEdges a set of extra edges we want to IGNORE during the search
+    * @param routeParams  a set of parameters that can restrict the route search
     * @return the computed route to the destination @targetNodeId
     */
-  def findRoute(g: DirectedGraph, localNodeId: PublicKey, targetNodeId: PublicKey, amountMsat: Long, numRoutes: Int, extraEdges: Set[GraphEdge] = Set.empty, ignoredEdges: Set[ChannelDesc] = Set.empty): Try[Seq[Hop]] = Try {
+  def findRoute(g: DirectedGraph,
+                localNodeId: PublicKey,
+                targetNodeId: PublicKey,
+                amountMsat: Long,
+                numRoutes: Int,
+                extraEdges: Set[GraphEdge] = Set.empty,
+                ignoredEdges: Set[ChannelDesc] = Set.empty,
+                routeParams: RouteParams): Try[Seq[Hop]] = Try {
+
     if (localNodeId == targetNodeId) throw CannotRouteToSelf
 
-    val foundRoutes = Graph.yenKshortestPaths(g, localNodeId, targetNodeId, amountMsat, ignoredEdges, extraEdges, numRoutes).toList match {
+    val boundaries: RichWeight => Boolean = { weight =>
+      ((weight.cost - amountMsat) < routeParams.maxFeeBaseMsat || (weight.cost - amountMsat) < (routeParams.maxFeePct * amountMsat)) &&
+      weight.length <= routeParams.routeMaxLength && weight.length <= DEFAULT_ROUTE_MAX_LENGTH &&
+      weight.cltv <= routeParams.routeMaxCltv
+    }
+
+    val foundRoutes = Graph.yenKshortestPaths(g, localNodeId, targetNodeId, amountMsat, ignoredEdges, extraEdges, numRoutes, boundaries).toList match {
+      case Nil if routeParams.routeMaxLength < DEFAULT_ROUTE_MAX_LENGTH =>
+        return findRoute(g, localNodeId, targetNodeId, amountMsat, numRoutes, extraEdges, ignoredEdges, routeParams.copy(routeMaxLength = DEFAULT_ROUTE_MAX_LENGTH, routeMaxCltv = DEFAULT_ROUTE_MAX_CLTV))
       case Nil => throw RouteNotFound
-      case route :: Nil if route.path.isEmpty => throw RouteNotFound
       case routes => routes.find(_.path.size == 1) match {
         case Some(directRoute) => directRoute :: Nil
         case _ => routes
       }
     }
 
-    // minimum cost
-    val minimumCost = foundRoutes.head.weight
-
-    // routes paying at most minimumCost + 10%
-    val eligibleRoutes = foundRoutes.filter(_.weight <= (minimumCost + minimumCost * DEFAULT_ALLOWED_SPREAD).round)
-    Random.shuffle(eligibleRoutes).head.path.map(graphEdgeToHop)
+    Random.shuffle(foundRoutes).head.path.map(graphEdgeToHop)
```

Comment:
this deserves a comment "at this point we know that foundRoutes is not empty".

Reply:
Done 582f2f7